File: apps/ios/dakka/Core/Call/CallManager.swift
import Foundation
import Combine
import WebRTC
import AVFoundation
import UIKit

/// Высокоуровневая оркестрация звонков:
/// • WebRTCManager (peer connection)
/// • SocketClient (signaling)
/// • CallStore (UI state)
/// • permissions (mic + camera)
@MainActor
final class CallManager: ObservableObject {
    static let shared = CallManager()

    private var rtc: WebRTCManager?
    private var pendingRemoteCandidates: [RTCIceCandidate] = []
    private var hasRemoteDescription = false
    private var cancellables = Set<AnyCancellable>()

    /// Кэш ICE-серверов от бэка (STUN + TURN).
    private var cachedICEServers: [RTCIceServer]?
    /// Fallback пока кэш не подгружен.
    private static let fallbackICE: [RTCIceServer] = [
        RTCIceServer(urlStrings: ["stun:stun.l.google.com:19302"]),
    ]

    private let store = CallStore.shared

    private init() {
        wireSocket()
        // Префетчим ICE-серверы при инициализации — к моменту первого звонка
        // в кэше уже будут TURN credentials.
        Task { await refreshICEServers() }
    }

    /// Обновляет кэш ICE-серверов с бэка.
    private func refreshICEServers() async {
        cachedICEServers = await ICEServerService.fetch()
    }

    // MARK: - Public API (выбор из UI)

    /// Инициатор звонка.
    func startCall(to peer: ChatMember, in chat: Chat, type: CallType) {
        guard case .idle = store.state else { return }
        let callId = "\(Int(Date().timeIntervalSince1970 * 1000))-\(UUID().uuidString.prefix(6))"
        let info = CallInfo(
            id: callId,
            chatId: chat.id,
            peerId: peer.userId,
            peerName: peer.user.displayName,
            peerAvatar: peer.user.avatar,
            type: type,
            isInitiator: true,
            startedAt: Date()
        )
        store.state = .outgoing(info)
        // CallKit: исходящий звонок в систему (для Recents и audio session)
        CallKitProvider.shared.startOutgoing(
            callId: callId, calleeName: peer.user.displayName, hasVideo: type == .video
        )
        SocketClient.shared.initiateCall(
            callId: callId, peerId: peer.userId, chatId: chat.id, callType: type
        )
        // Поднимаем WebRTC заранее — чтобы при accept сразу обмениваться offer
        setupRTC(isVideo: type == .video)
        sendOfferAfterPermission()
    }

    func acceptIncoming() {
        guard case .incoming(let info) = store.state else { return }
        Task {
            let ok = await requestPermissions(video: info.type == .video)
            guard ok else { endCurrent(reason: "no-permission"); return }
            setupRTC(isVideo: info.type == .video)
            SocketClient.shared.acceptCall(callId: info.id)
            store.state = .active(info)
            // Каллер пришлёт нам offer — мы ответим answer'ом в handleSignal()
        }
    }

    func rejectIncoming() {
        guard case .incoming(let info) = store.state else { return }
        SocketClient.shared.rejectCall(callId: info.id)
        CallKitProvider.shared.reportRemoteEnded(callId: info.id, reason: .declinedElsewhere)
        teardown()
    }

    func endCurrent(reason: String = "ended") {
        if let info = store.state.info {
            SocketClient.shared.endCall(callId: info.id)
        }
        teardown()
    }

    // MARK: - Controls

    func toggleMute() {
        store.isMuted.toggle()
        rtc?.isMuted = store.isMuted
    }

    func toggleVideo() {
        store.isVideoOff.toggle()
        rtc?.isVideoOff = store.isVideoOff
    }

    func toggleSpeaker() {
        store.isSpeakerOn.toggle()
        rtc?.setSpeakerOn(store.isSpeakerOn)
    }

    func switchCamera() {
        rtc?.switchCamera()
    }

    var localVideoTrack:  RTCVideoTrack? { rtc?.localVideoTrack }
    var remoteVideoTrack: RTCVideoTrack? { rtc?.remoteVideoTrack }

    // MARK: - Setup WebRTC

    private func setupRTC(isVideo: Bool) {
        guard rtc == nil else { return }
        // Создаём с STUN-fallback, затем асинхронно подгружаем актуальные
        // серверы и пересоздаём peer connection если получили TURN.
        // На практике setupRTC вызывается до createOffer/answer, так что
        // успеваем подтянуть до обмена ICE кандидатами.
        let m = WebRTCManager(
            isVideo: isVideo,
            iceServers: cachedICEServers ?? Self.fallbackICE
        )

        m.onIceCandidate = { [weak self] candidate in
            guard let self, let info = self.store.state.info else { return }
            let signal = CallSignal(
                type: "candidate",
                sdp: nil,
                candidate: candidate.sdp,
                sdpMid: candidate.sdpMid,
                sdpMLineIndex: Int32(candidate.sdpMLineIndex)
            )
            SocketClient.shared.sendCallSignal(callId: info.id, signal: signal)
        }
        m.onRemoteVideoTrack = { [weak self] _ in
            // SwiftUI прочитает rtc?.remoteVideoTrack через CallManager в ActiveCallView
            Task { @MainActor in self?.objectWillChange.send() }
        }
        m.onConnectionState = { [weak self] state in
            Task { @MainActor in
                self?.store.iceState = "\(state.rawValue)"
                // При соединении переключаем в .active (если ещё в outgoing)
                if state == .connected || state == .completed,
                   case .outgoing(let info) = self?.store.state ?? .idle {
                    self?.store.state = .active(info)
                    CallKitProvider.shared.reportConnected(callId: info.id)
                }
            }
        }
        self.rtc = m
    }

    private func sendOfferAfterPermission() {
        Task {
            let ok = await requestPermissions(video: store.state.info?.type == .video)
            guard ok, let info = store.state.info else { endCurrent(reason: "no-permission"); return }
            rtc?.createOffer { [weak self] sdp in
                guard let self, let sdp else { return }
                let signal = CallSignal(
                    type: "offer", sdp: sdp.sdp,
                    candidate: nil, sdpMid: nil, sdpMLineIndex: nil
                )
                SocketClient.shared.sendCallSignal(callId: info.id, signal: signal)
            }
        }
    }

    // MARK: - Signaling

    private func wireSocket() {
        SocketClient.shared.callIncoming
            .sink { [weak self] event in self?.handleIncoming(event) }
            .store(in: &cancellables)

        SocketClient.shared.callAccepted
            .sink { [weak self] event in self?.handleAccepted(event) }
            .store(in: &cancellables)

        SocketClient.shared.callEnded
            .sink { [weak self] event in self?.handleEnded(event) }
            .store(in: &cancellables)

        SocketClient.shared.callSignal
            .sink { [weak self] event in self?.handleSignal(event) }
            .store(in: &cancellables)
    }

    private func handleIncoming(_ event: IncomingCallEvent) {
        guard case .idle = store.state else {
            // Уже в звонке → busy
            SocketClient.shared.rejectCall(callId: event.callId)
            return
        }
        let info = CallInfo(
            id: event.callId,
            chatId: event.chatId,
            peerId: event.callerId,
            peerName: event.callerName,
            peerAvatar: event.callerAvatar,
            type: event.callType,
            isInitiator: false,
            startedAt: Date()
        )
        store.state = .incoming(info)
    }

    private func handleAccepted(_ event: CallAcceptedEvent) {
        // Каллер: другой принял, можно ждать когда ICE установится
        guard case .outgoing(let info) = store.state, info.id == event.callId else { return }
        // state переключится в .active по ICE connected (см. onConnectionState)
    }

    private func handleEnded(_ event: CallEndedEvent) {
        guard let info = store.state.info, info.id == event.callId else { return }
        CallKitProvider.shared.reportRemoteEnded(callId: info.id)
        teardown()
    }

    private func handleSignal(_ event: CallSignalEvent) {
        guard let info = store.state.info, info.id == event.callId, let rtc = self.rtc else { return }
        let signal = event.signal

        if signal.type == "offer", let sdp = signal.sdp {
            let desc = RTCSessionDescription(type: .offer, sdp: sdp)
            rtc.setRemoteDescription(desc) { [weak self] _ in
                self?.hasRemoteDescription = true
                self?.flushPendingCandidates()
                rtc.createAnswer { ans in
                    guard let ans else { return }
                    let signal = CallSignal(
                        type: "answer", sdp: ans.sdp,
                        candidate: nil, sdpMid: nil, sdpMLineIndex: nil
                    )
                    SocketClient.shared.sendCallSignal(callId: info.id, signal: signal)
                }
            }
        } else if signal.type == "answer", let sdp = signal.sdp {
            let desc = RTCSessionDescription(type: .answer, sdp: sdp)
            rtc.setRemoteDescription(desc) { [weak self] _ in
                self?.hasRemoteDescription = true
                self?.flushPendingCandidates()
            }
        } else if signal.type == "candidate", let cand = signal.candidate {
            let ice = RTCIceCandidate(
                sdp: cand,
                sdpMLineIndex: signal.sdpMLineIndex ?? 0,
                sdpMid: signal.sdpMid
            )
            if hasRemoteDescription {
                rtc.addRemoteCandidate(ice)
            } else {
                pendingRemoteCandidates.append(ice)
            }
        }
    }

    private func flushPendingCandidates() {
        guard let rtc = self.rtc else { return }
        for ice in pendingRemoteCandidates {
            rtc.addRemoteCandidate(ice)
        }
        pendingRemoteCandidates.removeAll()
    }

    // MARK: - Permissions

    private func requestPermissions(video: Bool) async -> Bool {
        let mic = await withCheckedContinuation { cont in
            if #available(iOS 17, *) {
                AVAudioApplication.requestRecordPermission { cont.resume(returning: $0) }
            } else {
                AVAudioSession.sharedInstance().requestRecordPermission { cont.resume(returning: $0) }
            }
        }
        guard mic else { return false }
        if video {
            let cam = await AVCaptureDevice.requestAccess(for: .video)
            return cam
        }
        return true
    }

    // MARK: - Teardown

    private func teardown() {
        rtc?.close()
        rtc = nil
        pendingRemoteCandidates.removeAll()
        hasRemoteDescription = false
        store.state = .idle
        store.isMuted = false
        store.isVideoOff = false
        store.isSpeakerOn = true
        store.iceState = "—"
    }
}

File: apps/android/app/src/main/java/online/akkdmsg/dakka/call/CallManager.kt
package online.akkdmsg.dakka.call

import android.content.Context
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import online.akkdmsg.dakka.data.ChatMember
import online.akkdmsg.dakka.data.api.ICEServerService
import online.akkdmsg.dakka.data.api.SocketClient
import org.webrtc.EglBase
import org.webrtc.IceCandidate
import org.webrtc.PeerConnection
import org.webrtc.SessionDescription
import java.util.UUID

/**
 * Высокоуровневая оркестрация звонков: WebRTC + signaling + state.
 * Singleton, чтобы переживать смену экранов и навигацию.
 */
object CallManager {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private var rtc: WebRTCManager? = null
    private var pendingCandidates: MutableList<IceCandidate> = mutableListOf()
    private var hasRemoteDescription = false
    private var cachedIce: List<PeerConnection.IceServer>? = null

    /** Общий EGL-контекст для VideoView'ов — должен совпадать между local/remote. */
    var eglBase: EglBase? = null
        private set

    private var appContext: Context? = null

    /** Вызывается из Application.onCreate (или лениво из первого старта звонка). */
    fun init(ctx: Context) {
        if (appContext != null) return
        appContext = ctx.applicationContext
        wireSocket()
        scope.launch { cachedIce = ICEServerService.fetch() }
    }

    // ── Public API ───────────────────────────────────────────────────────────

    fun startCall(peer: ChatMember, chatId: String, type: CallType) {
        if (CallStore.state.value !is CallState.Idle) return
        val callId = "${System.currentTimeMillis()}-${UUID.randomUUID().toString().take(6)}"
        val info = CallInfo(
            id = callId,
            chatId = chatId,
            peerId = peer.userId,
            peerName = peer.user.displayName,
            peerAvatar = peer.user.avatar,
            type = type,
            isInitiator = true,
            startedAt = System.currentTimeMillis(),
        )
        CallStore.setState(CallState.Outgoing(info))
        SocketClient.initiateCall(callId, peer.userId, chatId, type)
        setupRTC(type == CallType.Video) {
            it.createOffer { sdp ->
                sdp ?: return@createOffer
                SocketClient.sendCallSignal(callId, CallSignal(type = "offer", sdp = sdp.description))
            }
        }
    }

    fun acceptIncoming() {
        val info = (CallStore.state.value as? CallState.Incoming)?.info ?: return
        SocketClient.acceptCall(info.id)
        CallStore.setState(CallState.Active(info))
        setupRTC(info.type == CallType.Video) {
            // Каллер пришлёт offer — мы ответим answer'ом в handleSignal()
        }
    }

    fun rejectIncoming() {
        val info = (CallStore.state.value as? CallState.Incoming)?.info ?: return
        SocketClient.rejectCall(info.id)
        teardown()
    }

    fun endCurrent() {
        CallStore.state.value.info?.let { SocketClient.endCall(it.id) }
        teardown()
    }

    fun toggleMute() {
        val v = !CallStore.isMuted.value
        CallStore.setMuted(v)
        rtc?.setMuted(v)
    }

    fun toggleVideo() {
        val v = !CallStore.isVideoOff.value
        CallStore.setVideoOff(v)
        rtc?.setVideoEnabled(!v)
    }

    fun toggleSpeaker() {
        val v = !CallStore.isSpeakerOn.value
        CallStore.setSpeakerOn(v)
        rtc?.setSpeaker(v)
    }

    fun switchCamera() {
        rtc?.switchCamera()
    }

    val localVideoTrack get() = rtc?.localVideoTrack
    val remoteVideoTrack get() = rtc?.remoteVideoTrack
    val rendererEgl: EglBase? get() = rtc?.eglBase

    // ── Setup ────────────────────────────────────────────────────────────────

    private fun setupRTC(isVideo: Boolean, then: (WebRTCManager) -> Unit) {
        if (rtc != null) { then(rtc!!); return }
        val ctx = appContext ?: return
        val ice = cachedIce ?: listOf(
            PeerConnection.IceServer.builder("stun:stun.l.google.com:19302").createIceServer()
        )
        val egl = EglBase.create().also { eglBase = it }
        val m = WebRTCManager(ctx, isVideo, ice, egl).apply {
            onIceCandidate = { c ->
                val info = CallStore.state.value.info ?: return@apply
                SocketClient.sendCallSignal(info.id, CallSignal(
                    type = "candidate",
                    candidate = c.sdp,
                    sdpMid = c.sdpMid,
                    sdpMLineIndex = c.sdpMLineIndex,
                ))
            }
            onConnectionStateChange = { state ->
                if ((state == PeerConnection.IceConnectionState.CONNECTED ||
                     state == PeerConnection.IceConnectionState.COMPLETED) &&
                    CallStore.state.value is CallState.Outgoing) {
                    val info = (CallStore.state.value as CallState.Outgoing).info
                    CallStore.setState(CallState.Active(info))
                }
            }
        }
        rtc = m
        then(m)
    }

    // ── Socket subscription ──────────────────────────────────────────────────

    private fun wireSocket() {
        scope.launch {
            SocketClient.callIncoming.collect { event ->
                if (CallStore.state.value !is CallState.Idle) {
                    SocketClient.rejectCall(event.callId); return@collect
                }
                CallStore.setState(CallState.Incoming(CallInfo(
                    id = event.callId,
                    chatId = event.chatId,
                    peerId = event.callerId,
                    peerName = event.callerName,
                    peerAvatar = event.callerAvatar,
                    type = event.callType,
                    isInitiator = false,
                    startedAt = System.currentTimeMillis(),
                )))
            }
        }
        scope.launch {
            SocketClient.callAccepted.collect { _ -> /* state переключится по ICE connected */ }
        }
        scope.launch {
            SocketClient.callEnded.collect { event ->
                if (CallStore.state.value.info?.id == event.callId) teardown()
            }
        }
        scope.launch {
            SocketClient.callSignal.collect { event ->
                val info = CallStore.state.value.info ?: return@collect
                if (info.id != event.callId) return@collect
                handleSignal(event.signal)
            }
        }
    }

    private fun handleSignal(signal: CallSignal) {
        val rtc = rtc ?: return
        when (signal.type) {
            "offer" -> {
                val sdp = signal.sdp ?: return
                rtc.setRemoteDescription(SessionDescription(SessionDescription.Type.OFFER, sdp)) {
                    hasRemoteDescription = true
                    flushPending()
                    rtc.createAnswer { ans ->
                        ans ?: return@createAnswer
                        val info = CallStore.state.value.info ?: return@createAnswer
                        SocketClient.sendCallSignal(info.id, CallSignal(
                            type = "answer", sdp = ans.description,
                        ))
                    }
                }
            }
            "answer" -> {
                val sdp = signal.sdp ?: return
                rtc.setRemoteDescription(SessionDescription(SessionDescription.Type.ANSWER, sdp)) {
                    hasRemoteDescription = true
                    flushPending()
                }
            }
            "candidate" -> {
                val cand = signal.candidate ?: return
                val ice = IceCandidate(signal.sdpMid ?: "", signal.sdpMLineIndex ?: 0, cand)
                if (hasRemoteDescription) rtc.addRemoteCandidate(ice)
                else pendingCandidates.add(ice)
            }
        }
    }

    private fun flushPending() {
        val rtc = rtc ?: return
        pendingCandidates.forEach(rtc::addRemoteCandidate)
        pendingCandidates.clear()
    }

    private fun teardown() {
        rtc?.close()
        rtc = null
        eglBase?.release()
        eglBase = null
        pendingCandidates.clear()
        hasRemoteDescription = false
        CallStore.reset()
    }
}
